### PR TITLE
Proto Changes for DatabricksModelsArtifactRepo

### PR DESCRIPTION
**Pull request by arjundc-db**
Signed-off-by: Arjun DCunha <arjun.dcunha@databricks.com>

## What changes are proposed in this pull request?
This PR contains the proto changes required in order to facilitate the addition of a new artifact repository for the MLflow model registry

## How is this patch tested?

The new services can be imported and utilized in the client without any trouble.

### Is this a user-facing change?

- [x ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/protos/service.proto [hunk 1]
```diff
@@ -987,18 +988,6 @@ message ListArtifacts {
   }
 }
 
-// Metadata of a single artifact file or directory.
```
Comment (dbczumar) [inline mlflow/protos/service.proto:990]: I'm not a big fan of this change. Third-party MLflow service developers will need to start carrying around & compiling an additional `common.proto` file in order to implement the MLflow service API spec.

**Comment by ankit-db** [comment]
I don't have a ton of context, but @arjundc-db could you explain why a common proto is needed?

File: mlflow/protos/service.proto (hunk 1 above)
Comment (arjundc-db) [reply mlflow/protos/service.proto:990]: That's fair, I had to make multiple small changes to accommodate this myself. 
I initially created this to avoid duplicating the FileInfo across the tracking service and registry, and as a placeholder for any future common protos. However, on further discussion, it seems this would cause more issues than solve.
Thank you for pointing this out, reverting this change.

File: generate-protos.sh [hunk 2]
```diff
@@ -15,17 +15,23 @@ protoc -I="$PROTOS" \
...
 NEW_SCALAPB="from .scalapb import scalapb_pb2 as scalapb_dot_scalapb__pb2"
-sed -i'.old' -e "s/$OLD_SCALAPB/$NEW_SCALAPB/g" "$PROTOS/databricks_pb2.py" "$PROTOS/service_pb2.py" "$PROTOS/model_registry_pb2.py" "$PROTOS/databricks_artifacts_pb2.py"
+sed -i'.old' -e "s/$OLD_SCALAPB/$NEW_SCALAPB/g" "$PROTOS/databricks_pb2.py" "$PROTOS/service_pb2.py" "$PROTOS/model_registry_pb2.py" "$PROTOS/databricks_artifacts_pb2.py" "$PROTOS/databricks_model_artifacts_pb2.py"
 
 OLD_DATABRICKS="import databricks_pb2 as databricks__pb2"
 NEW_DATABRICKS="from . import databricks_pb2 as databricks__pb2"
-sed -i'.old' -e "s/$OLD_DATABRICKS/$NEW_DATABRICKS/g" "$PROTOS/service_pb2.py" "$PROTOS/model_registry_pb2.py" "$PROTOS/databricks_artifacts_pb2.py"
+sed -i'.old' -e "s/$OLD_DATABRICKS/$NEW_DATABRICKS/g" "$PROTOS/service_pb2.py" "$PROTOS/model_registry_pb2.py" "$PROTOS/databricks_artifacts_pb2.py" "$PROTOS/databricks_model_artifacts_pb2.py"
+
+OLD_SERVICE="import service_pb2 as service__pb2"
```
Comment (ankit-db) [inline generate-protos.sh:29]: What does this do?

File: mlflow/protos/databricks_model_artifacts_pb2.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
+# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (ankit-db) [inline mlflow/protos/databricks_model_artifacts_pb2.py:1]: Is this autogenerated?

File: mlflow/protos/databricks_model_artifacts.proto [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  option (scalapb.message).extends = "com.databricks.rpc.RPC[$this.Response]";
+
+  // Name of the registered model
+  optional string name = 1 [(validate_required) = true];
+
+  // Model version number
+  optional string version = 2 [(validate_required) = true];
+
+  // Filter artifacts matching this path (a relative path from the model's root artifact directory).
+  optional string path = 3;
```
Comment (ankit-db) [inline mlflow/protos/databricks_model_artifacts.proto:77]: What is the default behavior if this parameter is left empty? Could you make that clear in the docs?

File: mlflow/protos/databricks_model_artifacts.proto [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  // Callers must subsequently check for the existence of the artifacts using the appropriate
+  // cloud storage APIs
+  rpc getModelVersionSignedDownloadUri (GetModelVersionSignedDownloadUri) returns (GetModelVersionSignedDownloadUri.Response) {
+    option (rpc) = {
+      endpoints: [{
+        method: "GET",
+        path: "/mlflow/model-versions/get-signed-download-uri"
+        since { major: 2, minor: 0 },
+      }],
+      visibility: PUBLIC_UNDOCUMENTED,
```
Comment (ankit-db) [inline mlflow/protos/databricks_model_artifacts.proto:31]: re: visibility - is this temporary or the long term plan for this?

**Review by ankit-db — COMMENTED** [review]
Looks good to me - just some questions

File: mlflow/protos/databricks_model_artifacts_pb2.py (hunk 3 above)
Comment (arjundc-db) [reply mlflow/protos/databricks_model_artifacts_pb2.py:1]: Yep, running `generate-protos.sh` autogenerates all `*_pb2.py` files from the `*.proto` files.

File: mlflow/protos/databricks_model_artifacts.proto (hunk 4 above)
Comment (arjundc-db) [reply mlflow/protos/databricks_model_artifacts.proto:77]: This is optional since it is not required for the request to actually pass validation (be received by the MLflow service).
However, if an empty path is passed, then a 4xx code is returned saying 
`response: {"error_code":"INVALID_PARAMETER_VALUE","message":"Missing required field: path"}`

Comment (ankit-db) [reply mlflow/protos/databricks_model_artifacts.proto:77]: Got it - why not mark it as `validate_required` then?

File: generate-protos.sh (hunk 2 above)
Comment (arjundc-db) [reply generate-protos.sh:29]: Great question, I kinda took it for granted until you asked.
The essence of why we do this comes down to : https://github.com/protocolbuffers/protobuf/issues/1491
TLDR: https://github.com/protocolbuffers/protobuf/issues/1491#issuecomment-238123573
Thus we need to convert the relative imports into absolute import.

File: mlflow/protos/databricks_model_artifacts.proto (hunk 4 above)
Comment (arjundc-db) [reply mlflow/protos/databricks_model_artifacts.proto:77]: Makes sense, will do! Thanks

Comment (arjundc-db) [reply mlflow/protos/databricks_model_artifacts.proto:77]: Ohh wait, Sorry, I thought you were pointing towards the GetModelVersionSignedDownloadUri proto.
For list model artifacts, it is perfectly okay to not pass in the path, hence the optional. In fact, the `download_artifacts` will aim to list the artifacts without passing in the path.
When a call is made without a `path`, then the API returns a list of FileInfos from the root location ,i.e., from `dbfs:/databricks/mlflow-registry/<model-version-id>/models/<path-name>`.

**Comment by arjundc-db** [comment]
We decided to **not** make Proto changes to add the DatabricksModelsArtifactRepo.
